climada/engine/uncertainty/base.py

```diff
@@ -0,0 +1,685 @@
+        distributions using the chosen sampling_method from SALib.
+        https://salib.readthedocs.io/en/latest/api.html
+
+        Parameters
+        ----------
+        N : int
+            Number of samples as defined in SALib.sample.saltelli.sample().
+        sampling_method: string
+            The sampling method as defined in SALib. Possible choices:
+            'saltelli', 'fast_sampler', 'latin', 'morris', 'dgsm', 'ff'
+            https://salib.readthedocs.io/en/latest/api.html
+            The default is 'saltelli'
+        sampling_kwargs: dict()
+            Optional keyword arguments of the chosen SALib sampling method.
+
+        """
+        
+        uniform_base_sample = self._make_uniform_base_sample(N,
+                                                             sampling_method,
+                                                             sampling_kwargs)
+        df_samples = pd.DataFrame(uniform_base_sample, columns=self.param_labels)
+        for param in list(df_samples):
+            df_samples[param] = df_samples[param].apply(
+                self.distr_dict[param].ppf
+                )
+        self.sample = df_samples
+        LOGGER.info("Effective number of made samples: %d" %self.n_samples)
+        return df_samples
+
+
+    def _make_uniform_base_sample(self, N, sampling_method='saltelli',
+                                  sampling_kwargs=None):
+        """
+        Make a uniform distributed [0,1] sample for the defined
+        uncertainty parameters (self.param_labels) with the chosen
+        method from SALib (self.sampling_method)
+        https://salib.readthedocs.io/en/latest/api.html
+
+        Parameters
+        ----------
+        N:
+            Number of samples as defined for the SALib sample method.
+            Note that the effective number of created samples might be
+            larger (c.f. SALib)
+        sampling_method: string
+            The sampling method as defined in SALib. Possible choices:
+            'saltelli', 'fast_sampler', 'latin', 'morris', 'dgsm', 'ff'
+            https://salib.readthedocs.io/en/latest/api.html
+            The default is 'saltelli'
+        sampling_kwargs: dict()
+            Optional keyword arguments of the chosen SALib sampling method.
+
+        Returns
+        -------
+        sample_uniform : np.matrix
+            Returns a NumPy matrix containing the sampled uncertainty
+            parameters using the defined sampling method (self.sampling_method)
+
+        """
+        
+        if sampling_kwargs is None: sampling_kwargs = {}
+
+        #To import a submodule from a module use 'from_list' necessary
+        #c.f. https://stackoverflow.com/questions/2724260/why-does-pythons-import-require-fromlist
+        salib_sampling_method = getattr(
+            __import__('SALib.sample',
+                       fromlist=[sampling_method]
+                       ),
+            sampling_method
+            )
```

Comment:
I understand that the code is rather complex, although I think the name is pretty explicit and can be read even without understanding what is going on exactly. Your proposal would imho be very bad coding style. Even if it already exists in the CLIMADA codebase, it should not. 

Your proposal is to hard-code the dictionary of methods in the salib sample module. But this is exactly what the above lines of code are doing (extract the dictionnary of salib methods). This is much much cleaner, as 1) we do not introduce very hard to maintain redundant code 2) it can keep up with potential new releases (for example, if `salib` adds a new sampling function, this code would automatically work with it).

In my view, one should always avoid hard-coded variables if not necessary as these are basically impossible to maintain. 

The above code is a bit more advanced python, thus the link to the stackoverflow thread. If needed, one could add extra comments.

Reply:
hmm yes. i guess since one does not necessarily need to look at this beast when setting the kwarg sampling method in the function call, the advantages of having it like this may outweigh the hard-coding but easy to read option.

Reply:
I tried to slightly improve the comment in f296d47664e1a6d792462d0ca9f4587e0fcca62e